### ISSUE TITLE
Argument type `QueryT@QueryWrapper` does not satisfy upper bound `Query[_T@Query]` of type variable `Self`

### ISSUE THREAD

**Issue by pythonweb2** [astral-sh/ty#3871]
### Summary

I found something that seems like a false positive, and works with basedpyright

dependencies required: SQLAlchemy ~=2.0.51
```python
"""Minimal reproducer for ty type checker issue with Query|Select generic."""
from sqlalchemy.orm import Query
from sqlalchemy.sql import Select


class QueryWrapper[QueryT: Query | Select]:
    def __init__(self, value: QueryT) -> None:
        self._query = value

    @property
    def query(self) -> QueryT:
        return self._query

    @query.setter
    def query(self, value: QueryT) -> None:
        self._query = value

    def apply_filter(self, criterion) -> None:
        # ty complains about this
        self.query = self.query.filter(criterion)

        # but not this?
        if isinstance(self.query, (Query, Select)):
            self.query = self.query.filter(criterion)
```

The filter method is allowed on both a Select and Query object, and I don't really understand what the error message is trying to say.

### Version

ty 0.0.55

**Comment by sharkdp** [comment]
Thank you for reporting this. 

The full diagnostic output is:

```
error[invalid-argument-type]: Argument to bound method `Query.filter` is incorrect
  --> main.py:20:22
   |
20 |         self.query = self.query.filter(criterion)
   |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Argument type `QueryT@QueryWrapper` does not satisfy upper bound `Query[_T@Query]` of type variable `Self`
   |
info: Union variant `bound method QueryT@QueryWrapper.filter(*criterion: SQLCoreOperations[bool] | _HasClauseElement[bool] | TypedColumnsClauseRole[bool] | (() -> ColumnElement[bool]) | LambdaElement) -> QueryT@QueryWrapper` is incompatible with this call site
info: Attempted to call union type `(bound method QueryT@QueryWrapper.filter(*criterion: SQLCoreOperations[bool] | _HasClauseElement[bool] | TypedColumnsClauseRole[bool] | (() -> ColumnElement[bool]) | LambdaElement) -> QueryT@QueryWrapper) | (bound method QueryT@QueryWrapper.filter(*criteria: SQLCoreOperations[bool] | _HasClauseElement[bool] | TypedColumnsClauseRole[bool] | (() -> ColumnElement[bool]) | LambdaElement) -> QueryT@QueryWrapper)`

error[invalid-argument-type]: Argument to bound method `Select.filter` is incorrect
  --> main.py:20:22
   |
20 |         self.query = self.query.filter(criterion)
   |                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ Argument type `QueryT@QueryWrapper` does not satisfy upper bound `Select[_TP@Select]` of type variable `Self`
   |
info: Union variant `bound method QueryT@QueryWrapper.filter(*criteria: SQLCoreOperations[bool] | _HasClauseElement[bool] | TypedColumnsClauseRole[bool] | (() -> ColumnElement[bool]) | LambdaElement) -> QueryT@QueryWrapper` is incompatible with this call site
info: Attempted to call union type `(bound method QueryT@QueryWrapper.filter(*criterion: SQLCoreOperations[bool] | _HasClauseElement[bool] | TypedColumnsClauseRole[bool] | (() -> ColumnElement[bool]) | LambdaElement) -> QueryT@QueryWrapper) | (bound method QueryT@QueryWrapper.filter(*criteria: SQLCoreOperations[bool] | _HasClauseElement[bool] | TypedColumnsClauseRole[bool] | (() -> ColumnElement[bool]) | LambdaElement) -> QueryT@QueryWrapper)
```

I'm suspecting this is a duplicate of https://github.com/astral-sh/ty/issues/2585. I'll add it as a child of that issue.
